### PR TITLE
Issue #417: Retain whether or not a module was compiled with --count-instructions.

### LINKED ISSUE #417: Modules should record if instruction count instrumentation is present
Back when we merged support for instruction counts we didn't have an extensible way to record what optional features were enabled, and so the only way to know if the runtime can rely on instruction counting is seeing if the count is zero or not.

These days, we have exactly such a mechanism with [ModuleFeatures](https://github.com/bytecodealliance/lucet/blob/5b1b33bbe03e4f0405e6803b86f34309ee0afbe7/lucet-module/src/module_data.rs#L50-L62), so we should pass the existence (or lack of) a `--count-instructions` flag from lucetc to this struct for later reference.

### PR DESCRIPTION
**Pull request by fst-crenshaw**
This PR uses the `--count-instructions` option to set a new field in ModuleFeatures called `icnt`.  

**Notes:**
Leveraging the new field, the lucet-runtime call `get_instruction_count()` now returns an Option<u64>.  If the module was compiled with `--count-instructions`, the function returns the instruction count.  Otherwise, it returns None.

An extended set of tests affirm that get_instruction_count() appropriately returns Some(_) or None.

**See original issue at:**
https://github.com/bytecodealliance/lucet/issues/417

Fixes #417

### CODE REVIEW

File: lucet-module/src/module_data.rs [hunk 1]
```diff
@@ -58,6 +58,7 @@ pub struct ModuleFeatures {
     pub bmi2: bool,
     pub lzcnt: bool,
     pub popcnt: bool,
+    pub icnt: bool,
```
Comment (iximeow) [inline lucet-module/src/module_data.rs]: `lzcnt` and `popcnt` only have the abbreviated names to match [CPUID feature bits](https://en.wikipedia.org/wiki/Bit_Manipulation_Instruction_Sets#ABM) by the same name. I think you're entirely justified in using a more human friendly name like `instruction_count` here.

File: lucet-runtime/tests/instruction_counting.rs [hunk 2]
```diff
@@ -33,7 +32,40 @@ pub fn check_instruction_counts() {
             );
             ent
         })
-        .collect();
+        .collect()
+}
+
+#[test]
+pub fn check_icnt_off() {
```
Comment (iximeow) [inline lucet-runtime/tests/instruction_counting.rs]: :tada: effective test reuse!

File: lucet-module/src/module_data.rs (hunk 1 above)
Comment (fst-crenshaw) [reply lucet-module/src/module_data.rs]: Thanks for this feedback!  I'll change the name.

File: lucet-runtime/tests/instruction_counting.rs (hunk 2 above)
Comment (fst-crenshaw) [reply lucet-runtime/tests/instruction_counting.rs]: My thanks to you for writing a easy-to-extend test!

File: lucet-runtime/tests/instruction_counting.rs [hunk 3]
```diff
@@ -53,9 +90,12 @@ pub fn check_instruction_counts() {
         inst.run("test_function", &[]).expect("instance runs");
 
         let instruction_count = inst.get_instruction_count();
+        if let None = instruction_count {
+            panic!("instruction count expected from instance");
```
Comment (iximeow) [inline lucet-runtime/tests/instruction_counting.rs]: just a style thing: this `if` could be folded in with the `unwap` below into `instruction_count.expect("has instruction counting")`

Comment (fst-crenshaw) [reply lucet-runtime/tests/instruction_counting.rs]: Hm.  I attempted something like that but got a compiler message like "`expect` not available for u64`.  Possible problem between me and the keyboard.  I may be missing something.

File: lucet-runtime/tests/instruction_counting.rs [hunk 4]
```diff
@@ -33,7 +35,41 @@ pub fn check_instruction_counts() {
...
+        let region = MmapRegion::create(1, &Limits::default()).expect("region can be created");
+
+        let mut inst = region
+            .new_instance(module)
+            .expect("instance can be created");
+
+        inst.run("test_function", &[]).expect("instance runs");
+
+        let instruction_count = inst.get_instruction_count();
+        if let Some(_) = instruction_count {
```
Comment (katie-martin-fastly) [inline lucet-runtime/tests/instruction_counting.rs]: For situations where you are disregarding the inner value, (i.e. `Some(_)` or `Ok(_)`) there are also `is_some` and `is_none` methods! You can see more about `is_some` here: https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some

So, this condition could instead look like:

```suggestion
        if instruction_count.is_some() {
```

Comment (katie-martin-fastly) [reply lucet-runtime/tests/instruction_counting.rs]: this is getting into _Real Pedantic Hours_™ however, so this isn't something I'd consider blocking by any means.

Comment (katie-martin-fastly) [reply lucet-runtime/tests/instruction_counting.rs]: OOPS

This is an outdated diff, you've already fixed this.

**Pull request merged**